### PR TITLE
mgr/cephadm: Delete ceph.target if last cluster

### PR DESCRIPTION
**Pull request by rkachach**
Fixes: https://tracker.ceph.com/issues/46655

Signed-off-by: Redouane Kachach <rkachach@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/cephadm/cephadm [hunk 1]
```diff
@@ -6459,6 +6459,18 @@ def command_zap_osds(ctx: CephadmContext) -> None:
...
+def get_ceph_cluster_count(ctx):
+    # type: (CephadmContext) -> int
+    def is_valid_cluster_uuid(uuid_to_test):
+        # type: (str) -> bool
+        try:
+            uuid.UUID(str(uuid_to_test))
+            return True
+        except ValueError:
+            return False
+    return len([c for c in os.listdir(ctx.data_dir) if is_valid_cluster_uuid(c)])
```
Comment (mgfritch) [inline src/cephadm/cephadm:6471]:
```suggestion
def get_ceph_cluster_count(ctx: CephadmContext) -> int:
    return len([c for c in os.listdir(ctx.data_dir) if is_fsid(c)])
```

Comment (mgfritch) [reply src/cephadm/cephadm:6471]: why not use the existing `is_fsid` func? likewise, let's try to use py3 style type hints where possible ..

File: src/cephadm/cephadm [hunk 2]
```diff
@@ -6468,28 +6480,26 @@ def command_rm_cluster(ctx):
...
 
-    # stop + disable individual daemon units
-    for d in list_daemons(ctx, detail=False):
-        if d['fsid'] != ctx.fsid:
-            continue
-        if d['style'] != 'cephadm:v1':
-            continue
-        unit_name = get_unit_name(ctx.fsid, d['name'])
+    def disable_systemd_service(unit_name):
+        # type: (str) -> None
```
Comment (mgfritch) [inline src/cephadm/cephadm:6484]:
```suggestion
    def disable_systemd_service(unit_name: str) -> None:
```

File: src/cephadm/cephadm [hunk 3]
```diff
@@ -6519,7 +6529,9 @@ def command_rm_cluster(ctx):
     call_throws(ctx, ['rm', '-f', ctx.logrotate_dir + '/ceph-%s' % ctx.fsid])
 
     # if last cluster on host remove shared files
-    if not os.listdir(ctx.data_dir):
+    if get_ceph_cluster_count(ctx) == 0:
+
+        disable_systemd_service('ceph.target')
```
Comment (mgfritch) [inline src/cephadm/cephadm:6534]:
```suggestion
    if get_ceph_cluster_count(ctx) == 0:
        disable_systemd_service('ceph.target')
```
(nit)

File: src/cephadm/cephadm (hunk 1 above)
Comment (rkachach) [reply src/cephadm/cephadm:6471]: thanks @mgfritch, yeah it does the same, I didn't see that function :+1:

**Review by mgfritch — APPROVED** [review]

**Comment by adk3798** [comment]
http://pulpito.ceph.com/adking-2022-02-17_13:37:32-orch:cephadm-wip-adk-testing-2022-02-17-0053-distro-basic-smithi/

Failures tracked by https://tracker.ceph.com/issues/54304

**Pull request merged**